### PR TITLE
RFC: more generic convergence assessment

### PR DESCRIPTION
**Pull request by jonathanBieler**
I've tried to implement some of the changes discussed in #309, which allows my two minimal examples of [ZerothOrderOptimizer](https://gist.github.com/jonathanBieler/47d9ae7e95e7ca0f7352de8f84827ae3) and [FirstOrderOptimizer](https://gist.github.com/jonathanBieler/ed2ae8868e7b317c9e6d2db86f6ed2b9) defined outside of Optim to run.

### CODE REVIEW

**Comment by pkofod** [comment]
I love that the commits have two versions of you as authors :)

I like the change. Does this pass all tests locally @ you?

**Comment by jonathanBieler** [comment]
I don't know why there's two me there... tests are passing yes, I had a small mistake in the PR.

File: src/multivariate/optimize/optimize.jl [hunk 1]
```diff
@@ -13,6 +13,14 @@ update_h!(d, state, method::SecondOrderOptimizer) = hessian!(d, state.x)
 
 after_while!(d, state, method, options) = nothing
 
+initial_convergence(d, state, method, initial_x, options) = false 
+initial_convergence(d, state, method::ZerothOrderOptimizer, initial_x, options) = false
```
Comment (anriseth) [inline src/multivariate/optimize/optimize.jl]: Is this used anywhere?

File: src/utilities/assess_convergence.jl [hunk 2]
```diff
@@ -63,11 +63,13 @@ function default_convergence_assessment(state::AbstractOptimizerState, d, option
...
-
+    g_converged = gradient_convergence_assessment(state,d,options)
+    
     converged = x_converged || f_converged || g_converged
 
     return x_converged, f_converged, g_converged, converged, f_increased
 end
+
+gradient_convergence_assessment(state::AbstractOptimizerState, d, options) = g_residual(gradient(d)) ≤ options.g_tol
+gradient_convergence_assessment(state::ZerothOrderState, d, options) = false
```
Comment (anriseth) [inline src/utilities/assess_convergence.jl]: Is this used anywhere?

Comment (pkofod) [reply src/utilities/assess_convergence.jl]: Yes, in Jonathan's solver outside of Julia :) He wants to make some ZerothOrderSolver's outside of Optim but hook into everything else in Optim

**Comment by pkofod** [comment]
@jonathanBieler could you add a few tests? Just to make sure everything works as expected.

**Comment by jonathanBieler** [comment]
I've added a couple of tests, I had to make DummyMethod type and change a few things, since those were not used anywhere, I assumed it's fine.

**Pull request merged**